internal/tuf/v02/root.go

```diff
@@ -0,0 +1,382 @@
+func (r *RootMetadata) DeletePrimaryRuleFilePrincipal(principalID string) error {
+	if principalID == "" {
+		return tuf.ErrInvalidPrincipalID
+	}
+
+	targetsRole, ok := r.Roles[tuf.TargetsRoleName]
+	if !ok {
+		return tuf.ErrPrimaryRuleFileInformationNotFoundInRoot
+	}
+
+	if targetsRole.PrincipalIDs.Len() <= targetsRole.Threshold {
+		return tuf.ErrCannotMeetThreshold
+	}
+
+	targetsRole.PrincipalIDs.Remove(principalID)
+	r.Roles[tuf.TargetsRoleName] = targetsRole
+	return nil
+}
+
+// AddGitHubAppPrincipal adds the 'principal' as a trusted principal in
+// 'rootMetadata' for the special GitHub app role. This key is used to verify
+// GitHub pull request approval attestation signatures.
+func (r *RootMetadata) AddGitHubAppPrincipal(principal tuf.Principal) error {
+	if principal == nil {
+		return tuf.ErrInvalidPrincipalType
+	}
+
+	// TODO: support multiple principals / threshold for app
+	if err := r.addPrincipal(principal); err != nil {
+		return err
+	}
+	role := Role{
+		PrincipalIDs: set.NewSetFromItems(principal.ID()),
+		Threshold:    1,
+	}
+	r.addRole(tuf.GitHubAppRoleName, role) // AddRole replaces the specified role if it already exists
+	return nil
+}
+
+// DeleteGitHubAppPrincipal removes the special GitHub app role from the root
+// metadata.
+func (r *RootMetadata) DeleteGitHubAppPrincipal() {
+	// TODO: support multiple principals / threshold for app
+	delete(r.Roles, tuf.GitHubAppRoleName)
+}
+
+// EnableGitHubAppApprovals sets GitHubApprovalsTrusted to true in the
+// root metadata.
+func (r *RootMetadata) EnableGitHubAppApprovals() {
+	r.GitHubApprovalsTrusted = true
+}
+
+// DisableGitHubAppApprovals sets GitHubApprovalsTrusted to false in the root
+// metadata.
+func (r *RootMetadata) DisableGitHubAppApprovals() {
+	r.GitHubApprovalsTrusted = false
+}
+
+// UpdateRootThreshold sets the threshold for the Root role.
+func (r *RootMetadata) UpdateRootThreshold(threshold int) error {
+	rootRole, ok := r.Roles[tuf.RootRoleName]
+	if !ok {
+		return tuf.ErrInvalidRootMetadata
+	}
+
+	if rootRole.PrincipalIDs.Len() < threshold {
+		return tuf.ErrCannotMeetThreshold
+	}
+	rootRole.Threshold = threshold
+	r.Roles[tuf.RootRoleName] = rootRole
+	return nil
+}
+
+// UpdatePrimaryRuleFileThreshold sets the threshold for the top level Targets
+// role.
+func (r *RootMetadata) UpdatePrimaryRuleFileThreshold(threshold int) error {
+	targetsRole, ok := r.Roles[tuf.TargetsRoleName]
+	if !ok {
+		return tuf.ErrPrimaryRuleFileInformationNotFoundInRoot
+	}
+
+	if targetsRole.PrincipalIDs.Len() < threshold {
+		return tuf.ErrCannotMeetThreshold
+	}
+	targetsRole.Threshold = threshold
+	r.Roles[tuf.TargetsRoleName] = targetsRole
+	return nil
+}
+
+// GetPrincipals returns all the principals in the root metadata.
+func (r *RootMetadata) GetPrincipals() map[string]tuf.Principal {
+	principals := map[string]tuf.Principal{}
+	for id, key := range r.Keys {
+		principals[id] = key
+	}
+	for id, person := range r.Persons {
+		principals[id] = person
```

Comment:
principal is a generic interface for some entity that can sign / approve something. It can be a single key, it can be a person, etc. The metadata uses the generic terminology so we can support all of these in parallel.